crates/ide/src/doc_links.rs

```diff
@@ -308,19 +322,38 @@ fn broken_link_clone_cb(link: BrokenLink<'_>) -> Option<(CowStr<'_>, CowStr<'_>)
 //
 // This should cease to be a problem if RFC2988 (Stable Rustdoc URLs) is implemented
 // https://github.com/rust-lang/rfcs/pull/2988
-fn get_doc_link(db: &RootDatabase, def: Definition) -> Option<String> {
-    let (target, file, frag) = filename_and_frag_for_def(db, def)?;
+fn get_doc_links(
+    db: &RootDatabase,
+    def: Definition,
+    target_dir: Option<&OsStr>,
+) -> DocumentationLinks {
+    let Some((target, file, frag)) = filename_and_frag_for_def(db, def) else { return Default::default(); };
 
-    let mut url = get_doc_base_url(db, target)?;
+    let (mut web_url, mut local_url) = get_doc_base_urls(db, target, target_dir);
 
     if let Some(path) = mod_path_of_def(db, target) {
-        url = url.join(&path).ok()?;
+        web_url = join_url(web_url, &path);
+        local_url = join_url(local_url, &path);
     }
 
-    url = url.join(&file).ok()?;
-    url.set_fragment(frag.as_deref());
+    web_url = join_url(web_url, &file);
+    local_url = join_url(local_url, &file);
+
+    set_fragment_for_url(web_url.as_mut(), frag.as_deref());
+    set_fragment_for_url(local_url.as_mut(), frag.as_deref());
 
-    Some(url.into())
+    return DocumentationLinks {
+        web_url: web_url.map(|it| it.into()),
+        local_url: local_url.map(|it| it.into()),
+    };
+
+    fn join_url(base_url: Option<Url>, path: &str) -> Option<Url> {
+        base_url.and_then(|url| url.join(path).ok())
+    }
+
+    fn set_fragment_for_url(url: Option<&mut Url>, frag: Option<&str>) {
+        url.map(|url| url.set_fragment(frag));
+    }
```

Comment:
nit: let's make these two functions closures (there is an assist for that) and stick them at the top of the function, that way its less noisy (and we can get rid of the trailing return)